sklearn/preprocessing/_polynomial.py

```diff
@@ -1003,33 +1063,53 @@ def transform(self, X):
 
                 if use_sparse:
-                    XBS_sparse = BSpline.design_matrix(
-                        x, spl.t, spl.k, **kwargs_extrapolate
-                    )
-                    if self.extrapolation == "periodic":
-                        # See the construction of coef in fit. We need to add the last
-                        # degree spline basis function to the first degree ones and
-                        # then drop the last ones.
-                        # Note: See comment about SparseEfficiencyWarning below.
-                        XBS_sparse = XBS_sparse.tolil()
-                        XBS_sparse[:, :degree] += XBS_sparse[:, -degree:]
-                        XBS_sparse = XBS_sparse[:, :-degree]
+                    x = x.copy()  # copy to avoid inplace operation
+                    # we leave the nan values out of bsplining and make them into 0s
+                    # right away, since BSpline.design_matrix() would raise on the nans
+                    # when X is sparse:
+                    design_matrix = [
+                        (
+                            self._make_design_matrix(
+                                ele,
+                                spl,
+                                self.extrapolation,
+                                degree,
+                                **kwargs_extrapolate,
+                            )
+                            if not np.isnan(ele)
+                            else [0] * n_splines
+                        )
+                        for ele in x
+                    ]
+                    design_matrix_csr = [
+                        sparse.csr_matrix(arr) for arr in design_matrix
+                    ]
+                    XBS_sparse = sparse.vstack(design_matrix_csr)
+
                 else:
                     XBS[:, (i * n_splines) : ((i + 1) * n_splines)] = spl(x)
+                    # replace any indicated values with 0:
+                    extended_nan_indicator = np.repeat(nan_indicator, n_splines, axis=1)
+                    XBS[extended_nan_indicator] = 0
+                    if self.sparse_output:
+                        XBS = sparse.csr_matrix(XBS)
+
             else:  # extrapolation in ("constant", "linear")
                 xmin, xmax = spl.t[degree], spl.t[-degree - 1]
                 # spline values at boundaries
                 f_min, f_max = spl(xmin), spl(xmax)
+                # values outside of the feature space during `fit` and nan values get
+                # masked out:
                 mask = (xmin <= X[:, i]) & (X[:, i] <= xmax)
+
                 if use_sparse:
                     mask_inv = ~mask
                     x = X[:, i].copy()
-                    # Set some arbitrary values outside boundary that will be reassigned
-                    # later.
+                    # Set some arbitrary values that will be reassigned later:
                     x[mask_inv] = spl.t[self.degree]
```

Comment:
@lorentzenchr, I'd have a question here: What is the reason you had set these arbitrary values to `spl.t[self.degree]`? I cannot see why this value would always be within the feature space. And: can I change it to `xmin` or so?